common/determined_common/schemas/extensions.py

```diff
@@ -0,0 +1,159 @@
+            if len(instance) != 1:
+                return False
+            return name in instance
+        if key.startswith("type:"):
+            typ = key[len("type:") :]
+            assert typ in ["array", "object"]
+            if typ == "array":
+                return isinstance(instance, list)
+            if typ == "object":
+                return isinstance(instance, dict)
+        if key.startswith("hasattr:"):
+            attr = key[len("hasattr:") :]
+            return isinstance(attr, dict) and attr in instance
+    raise ValueError(f"invalid unionKey: {key}")
+
+
+def union(validator, det_one_of, instance, schema):
+    # like oneOf but with custom error handling
+    message = det_one_of.get("defaultMessage", "union failed to validate")
+
+    selected_errors = None
+    valid = []
+
+    for idx, item in enumerate(det_one_of["items"]):
+        errors = [e for e in validator.descend(instance, schema=item, schema_path=idx)]
+        if errors:
+            key = item["unionKey"]
+            if not selected_errors and _evaluate_unionKey(key, instance):
+                selected_errors = errors
+        else:
+            valid.append(item)
+
+    if len(valid) == 1:
+        # No errors.
+        return
+
+    if len(valid) > 1:
+        yield jsonschema.ValidationError(
+            f"bug in validation! Multiple schemas matched: {valid}"
+        )
+        return
+
+    if selected_errors:
+        yield from selected_errors
+        return
+
+    yield jsonschema.ValidationError(
+        det_one_of.get("defaultMessage", "union failed to validate")
+    )
+
+
+def disallowProperties(validator, disallowed, instance, schema):
+    if not validator.is_type(instance, "object"):
+        return
+
+    for prop in instance:
+        if prop in disallowed:
+            msg = disallowed[prop]
+            yield jsonschema.ValidationError(msg)
+
+
+def compareProperties(validator, compare, instance, schema):
+    if not validator.is_type(instance, "object"):
+        return
+
+    def get(prop):
+        obj = instance
+        for key in prop.split("."):
+            if not obj:
+                return None
+            obj = obj.get(key)
+        return obj
+
+    a_prop = compare["a"]
+    a = get(a_prop)
+
+    b_prop = compare["b"]
+    b = get(b_prop)
+
+    if a is None or b is None:
+        return
+
+    typ = compare["type"]
+
+    if typ == "a<b":
+        if a >= b:
+            yield jsonschema.ValidationError(f"{a_prop} must be less than {b_prop}")
+        return
+
+    if typ == "same_units":
+        if not isinstance(a, dict) or not isinstance(b, dict):
+            return
+        if set(a.keys()) != set(b.keys()):
+            yield jsonschema.ValidationError(f"{a_prop} must be defined in the same units as {b_prop}")
+
+    if typ == "length_a<length_b":
+        if not isinstance(a, dict) or isinstance(b, dict):
+            return
+        # Assume the same units.
+        length_a = next(a.items())
```

Comment:
I remember reading the go parallel to this and it looks so much more ok exactly translated in python, ha.

Reply:
double lol since this line turned out to have two bugs: `a.items()` should be `a.values()`, and you need to `next(iter())` not just `next()`.  I added a bunch more test cases.